x-pack/apm-server/sampling/processor.go

```diff
@@ -144,24 +145,29 @@ func (p *Processor) ProcessBatch(ctx context.Context, batch *model.Batch) error
 	for i := 0; i < len(events); i++ {
 		event := &events[i]
 		var report, stored bool
+		var err error
 		switch event.Processor {
 		case model.TransactionProcessor:
-			var err error
 			atomic.AddInt64(&p.eventMetrics.processed, 1)
 			report, stored, err = p.processTransaction(event)
-			if err != nil {
-				return err
-			}
 		case model.SpanProcessor:
-			var err error
 			atomic.AddInt64(&p.eventMetrics.processed, 1)
 			report, stored, err = p.processSpan(event)
-			if err != nil {
-				return err
-			}
 		default:
 			continue
 		}
+
+		// If processing the transaction or span returns with an error and
+		// it's caused by the OS returning a "no space left on device", we
+		// sample the transaction or span by default.
+		// NOTE(marclop) This behavior should be configurable so users who
+		// rely on tail sampling for cost cutting, can discard events once
+		// the disk is full.
+		if err != nil && ignoreENOSPCErr(err) == nil {
+			p.logger.Info("received disk is full error, sampling trace by default")
```

Comment:
Might be a good idea to capture another metric for this (count of write failures), updating it in updateProcessorMetrics